Sally7/Internal/WireFormatting.cs

```diff
@@ -0,0 +1,67 @@
+﻿using Sally7.Protocol.S7.Messages;
+using Sally7.RequestExecutor;
+
+namespace Sally7.Internal;
+
+internal static class WireFormatting
+{
+    private const int Tpkt = 0x03_00_00_00;
+
+    /// <summary>
+    /// Data struct.
+    /// </summary>
+    /// <remarks>
+    /// This struct is only 3 bytes long, final byte is overwritten.
+    /// </remarks>
+    private const int Data = 2 << 24 // Length
+        | 0b1111_0000 << 16 // Identifier
+        | 0b1000_0000 << 8; // PDU number and EOT
+
+    private const int JobRequestHeader1 = 0x32 << 24 // Protocol ID
+        | (byte)MessageType.JobRequest << 16; // Message type
+
+    /// <summary>
+    /// The PDU reference.
+    /// </summary>
+    /// <remarks>
+    /// This value is here for legacy reasons. The <see cref="ConcurrentRequestExecutor"/> modifies the lower
+    /// byte to identify responses, the upper byte isn't actually used at the moment.
+    /// </remarks>
+    private const short PduRef = 1 << 8;
+
+    public static int WriteTpkt(ref byte destination, int length)
+    {
+        return WriteInt32(ref destination, Tpkt | length);
+    }
+
+    public static int WriteData(ref byte destination)
+    {
+        // For performance we write an int instead of separate values, the final byte of the int will be overwritten.
+        WriteInt32(ref destination, Data);
+
+        return 3;
+    }
+
+    public static int WriteJobRequestHeader(ref byte destination, int paramLength, int dataLength)
+    {
+        WriteInt32(ref destination, JobRequestHeader1);
+        // Legacy, the
+        WriteInt16(ref destination.GetOffset(4), 1 << 8); // PDU ref
+        WriteInt16(ref destination.GetOffset(6), (short)paramLength);
+        WriteInt16(ref destination.GetOffset(8), (short)dataLength);
```

Comment:
This code is nice and very readable.

Perf-wise we have 4 memory writes -- which may or may not be coalesced by cpu's store buffer.
So maybe combine these writes into a `long` + `short` (sure taking byte order into account), so we reduced that to 2 memory writes.

I'm not sure if a micro-benchmark will show any difference here. More likely a tool like Intel VTune can show differences in memory operations.

Further I believe that the code can be structed in a way that's still readable and understandable.
Maybe we should look into this in a follow-up PR.

Reply:
I'll add a benchmark. I also have a separate [repository for benchmarks](https://github.com/mycroes/Benchmarks), I'll see if I can put it into a usable form there as well. Either way I definitely agree, less writes is faster so it probably helps to combine it into a long + short write.

Reply:
Nice repo -- I'll dig into this next week 😃.

Reply:
So, lunch done and I gave it a try:
```c#
    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    public static int WriteJobRequestHeader(ref byte destination, int paramLength, int dataLength)
    {
        ulong tmp;

        if (BitConverter.IsLittleEndian)
        {
            tmp  = (ulong)BinaryPrimitives.ReverseEndianness(JobRequestHeader1);
            tmp |= (ulong)BinaryPrimitives.ReverseEndianness((short)(1 << 8)) << 32;
            tmp |= (ulong)(uint)BinaryPrimitives.ReverseEndianness((ushort)paramLength) << 48;
        }
        else
        {
            tmp  = JobRequestHeader1;
            tmp |= (ulong)(1 << 8) << 32;
            tmp |= (ulong)(uint)paramLength << 48;
        }

        Unsafe.WriteUnaligned(ref destination, tmp);

        WriteUInt16(ref destination.GetOffset(8), (ushort)dataLength);

        return 10;
    }
```
* code is not so pretty
* I'm not sure if the big-endian part is correct (this is where I'm always struggling)
* as this is the only place where this long+short is used, I don't think a helper is needed (similar to the `NetworkOrderSerializer`-class)

Maybe you have a better idea.
[Codegen-wise](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIAYACY8gOgCEBXAMy5il3YBLAHbYoATwDcNek1YAlDsIyD8MFgGEI+AA6CANnwDKfAG6CwMXNNqNmLRctXqAksr4QdJqOcvWaNCIYfKL6ckhMpAwA6oKwAGLQ+NgYKsIA5jQA3jQMeQw6UIKmKTAMkMK4GAxBDAAqOgDW1QC8DHQIdADMAPp0dH0D/Ta5+QD0EwwAPLgc+MkSAHyjeRNjDAAiKdgMVVAcYBgsKwxr02Oz82Liy9T5p5NTsAuNuLf3Z3UAFoK4uxj7Q41P4QYT6cQMLoMYDiYJ/fSg9JoBhcETYMIw4LAhgQUx8ADuRVSMGExzu40eY2eYle73yhWKpXKoKqNWUm22DDaUSmUwYpBQDwYABkSekMF8TvdCO1gOR5eRBv1pnzyBFJi4ACYklSovhS/Iyuhy/pDU0qhgADkkQoAChsAKoMYRzYB8BjYYSahgAUQA8nUAuS8gySliKqzagApCDAeQwACOHCsGAAEjBsNqoOQue0EF0eXyBXaoBAMBBIGEXBsDXkZQAKTEwACUAFkrLhsOkYHVxDp1DG44nk6zeQw1TbJu3cJ3uwwMH2YEGPo9LgsbidPl8yvanbBeLBhJYySv1lMLnN13TVo9qVBaZvJt9fgwSvpk9jt7AUdAGIZ0tgYAQrAeAsiw9TbtMuAwGUYD7i0ABEWhHhwUCHhg8ZJimPoIDAYAcOWUAIWMiwMPgECaoIep/BKZQIvi+rBg86xNvOEA1NqThcMBVg6CyVjIrRDAcDo/ZQNCsJlL8wgAOTVIBGAcOi4LCdB3opPOkHkWoygnhSZ5UjALxvCcoZMhG1S4F80DVLamocPGXC5jmY7Wsu+TEFCzARLU0REj2TQYPW+4SVi2pVGiKigsitSGBkErNicORMfcxAAOwxP5DpuBgBbBTATnhWkKSCNF9SBQwDYcEEzZxeKXzNjY9wAL7uXknnhGy1R+YIwRbBg2D5U5rFFZFpXCIlTHJfcK4MIk4liVwSSepYDAMWt/kesIXVslUGbehATnQToYhMm+I6CZBqKhKFZSHZpUnsviBgYndeJQISvXBKStaZb1MDZcoeUhaNohRcIyL9dgjVtTN6WQk1+StdQJwANrthKFEuLo+j1hj1matjOj6H6OjgwIACC6TpLAM7FDAbj6CIIjpM2AC6Jwdd5O09cEg6YSOab7XwQ23QwoMlWVtQnVA2D4KK8VfDF7KatsCv1ZN9zTTNeS8wDOXAwV4spmNZX88OKbppmfDkDDKUUiKMAAUBl1LvbutZTlaqixL4MsAA4jAGB+jw0FBSgzbIi5fKWo1dqOgw+6/XrgMYN7IMm2D40B0HIdcGH9ZIJHDD1hwVk2c2Mty+rCWIzrKde0gPuZ5LpKB8HodB/WsfIqX5dQBgzaqwNNcNSM7tMBl5B0HXDDI2j+NYzjeNBwTRMk2T42U9TtO4PTjPMxk7Oc15ao8/55tYVUVtZj0FOiyNLfg8r1RV/LYoSi/4tqx/DVJb9HAEQZHnLoceOs8hURLmwXqyF3rBCgCwFwuBhRfUMD6L0ghPSa3AdrcB+QMC6DyG0UuQDWbQNEBIW0RR8C9XpgIeM71oLoMop6YQHZ6yX0FjfG2ds8H3AIToSqxDAGImbOQ64VDVC0LxPQmAjCYDMMwcINhM56z1n7oPes0crTNmbBaAss88ECKESXERR9S41XEZQ6h0irAODkXwJhGDWHsL7tZAeldTrv0VnoscKA3IT2RngmA+hoK/VwXw4xuZOGW2FtmQx4DjGEGEaQ5sWiLSx30aQBJOskkpNERY5QnjZbePqhafxOS56w3uA6So2BeAsBTqEQQ6Q2GambhFLOZUBF22Tp7ZQ6cja+2zu3POBce6mI0UPH+PiwHgPhtPWe88mJczPr5fpacm4hUfp01uyIy7uOqOdFs/8J4ADkg74mgI0P0UAszeEwUzAAXnwRpGzBmFSfuNZExzeHzIynvF5h11GHL+XkZZqVT4+XZA3IGpAH6SWNrs5+wlai/NOXgi5GArn3lufcvgjzBAvIQbC3K8KM7Iu+a+dEyYwVwwBUSmAwKapLJoMsoIIR0SdTIPURcOFvp7xZNkE+nVtmItGZ3IKEoXxirCl8qW7JDr5yDtgvIETUoZRCrUzsDSKaanaRS4qKLCmDyVWHOlyN2XuCgDdbmPKsU4puXc4wBL0SMqgMKlZUKmCClJR8sWwyyoHJstS98Jypq/W1fU9QTS3WtJgAaoZ8qIa/XuNAjAsC+DwMQcg1BCjnHbQAPwMCseISRNCVAyPsfIxRLjVG/IYCAUNtLWUoy9dy31GzDbDURYGiGqL2ToojRPKNDTY1M3jYmz5lLoqpvyOmzNA9XlIJQakNBBaGDFtLeW2xsia0FpUbgesDam2/NbVa+BN0SRzAYNOWcPZFyNtup6+4MTWRtA6HQcgaATgUzAI0XMn7SA/qYn+xoUNAOdC6CBmp0EoAQY/Z0NKbKaBAA) it's only two memory writes at the expense of more register chaining. So hard to tell (w/o measuring) what's faster. In doubt I'd keep the code as is now, because with the NetworkOrderSerializer it's so beautiful to read.

Reply:
I guess this is hard to get your head around (still hurts my head as well), but reversing the endianess of the entire long suffices. It does make some sense, because for big endian all of the data is in the order that we want it on the network, so for little endian it's all in the incorrect order. That reduces the code complexity as well, and it's actually really fast (see benchmarks). Will definitely use long+short writes for this.